ipa-core/src/protocol/basics/share_validation.rs

```diff
@@ -0,0 +1,166 @@
+use futures_util::future::try_join;
+
+use crate::{
+    error::Error,
+    helpers::{
+        hashing::{compute_hash, Hash},
+        Direction,
+    },
+    protocol::{context::Context, RecordId},
+    secret_sharing::SharedValue,
+};
+
+/// This function checks that a vector of shares are consistent across helpers
+/// i.e. `H1` holds `(x0,x1)`, `H2` holds `(x1,x2)`, `H3` holds `(x2,x0)`
+/// the function verifies that `H1.x0 == H3.x0`, `H1.x1 == H2.x1`, `H2.x2 == H3.x2`
+///
+/// We use a hash based approach that is secure in the random oracle model
+/// further, only one of left and right helper check that it is zero
+/// this is might not be sufficient in some applications to prevent malicious behavior
+///
+/// The left helper simply hashes the vector and sends it to the right,
+/// the right helper negates his vector, hashes it and compares it to the received hash
+///
+/// # Errors
+/// propagates errors from send and receive
+pub async fn validate_replicated_shares<'a, 'b, C, I, J, S>(
+    ctx: C,
+    input_left: I,
+    input_right: J,
+) -> Result<(), Error>
+where
+    C: Context,
+    I: IntoIterator<Item = &'a S>,
+    J: IntoIterator<Item = &'b S>,
+    S: SharedValue,
+{
+    // compute hash of `left`
+    let hash_left = compute_hash(input_left);
+
+    // set up context
+    let ctx_new = &(ctx.set_total_records(1usize));
+    // set up channels
+    let send_channel = ctx_new.send_channel::<Hash>(ctx.role().peer(Direction::Right));
+    let receive_channel = ctx_new.recv_channel::<Hash>(ctx.role().peer(Direction::Left));
+
+    let ((), hash_right) = try_join(
+        // send hash
+        send_channel.send(RecordId::FIRST, compute_hash(input_right)),
+        receive_channel.receive(RecordId::FIRST),
+    )
+    .await?;
+
+    if hash_left == hash_right {
+        Ok(())
+    } else {
+        Err(Error::InconsistentShares)
+    }
+}
+
+/// This function is similar to validate the consistency of shares with the difference
+/// that it validates that the shares sum to zero rather than being identical
+/// i.e. `H1` holds `(x0,x1)`, `H2` holds `(x1,x2)`, `H3` holds `(x2,x0)`
+/// the function verifies that `H1.x0 == -H3.x0`, `H1.x1 == -H2.x1`, `H2.x2 == -H3.x2`
+///
+/// We use a hash based approach that is secure in the random oracle model
+/// further, only one of left and right helper check that it is zero
+/// this is sufficient for Distributed Zero Knowledge Proofs
+/// but might not be sufficient for other applications
+///
+/// The left helper simply hashes the vector and sends it to the right,
+/// the right helper negates his vector, hashes it and compares it to the received hash
+///
+/// # Errors
+/// propagates errors from `validate_replicated_shares`
+pub async fn validate_sum_to_zero<'a, C, I, S>(
```

Comment:
I think there is a big impedance mismatch between the name of this protocol, this comment, and how it is going to be used. 

It's easy to generate a secret-sharing of zero that would **fail** this test.

Let's generate the 3 shares like this:

x0 = rand()
x1 = rand()
x2 = PRIME - x0 - x1.

And then distribute them exactly as the comment above says:
`/// i.e. `H1` holds `(x0,x1)`, `H2` holds `(x1,x2)`, `H3` holds `(x2,x0)``

This function `validate_sum_to_zero` would error `Error::InconsistentShares`.

That's super confusing because it DOES "sum to zero", and when you look at the comment, you'd say: "But the shares are like that".

What this function is *actually* doing is checking that we have a sharing of the nature:
/// `H1` holds `(x0,-x1)`, `H2` holds `(x1,-x2)`, `H3` holds `(x2,-x0)`

So I'd suggest changing the comment to say *that* about how the shares are distributed, and then rename it to: `validate_three_two_way_sharings_of_zero` or similar.